### ISSUE TITLE
CoxTimeVaryingFitter() raising `ZeroDivisionError: float division by 0`

### ISSUE THREAD

**Issue by indiana-nikel** [CamDavidsonPilon/lifelines#768]
Hi there,

I'm looking to fit a simple time-varying model using `CoxTimeVaryingFitter()` using a cumulative sum of an event occurring. However, every attempt at using the `.fit()` method results in a `ZeroDivisionError: float division by zero`. I've attached a sample XLSX of what a few of the observations I'm using look like: 
[gh_issue.xlsx](https://github.com/CamDavidsonPilon/lifelines/files/3370042/gh_issue.xlsx).

The code for fitting looks like this:
```
import numpy as np
import pandas as pd
from lifelines import CoxTimeVaryingFitter

df_fit = pd.read_csv("gh_issue.csv")
ctv = CoxTimeVaryingFitter()
ctv.fit(df_fit, id_col="ID", event_col="Churn", start_col="Start", stop_col="Stop", show_progress=True)
```

The error appears to be with `weighted_average = weight_count / tied_death_counts` when calling `self._get_gradients()`. Is this intended behavior? If so, what is it signifying to the user?

Cheers,
Indiana

**Comment by CamDavidsonPilon** [comment]
Thanks for the detailed report! This is defs not intended. I'll investigate this tomorrow.

**Comment by indiana-nikel** [comment]
Wonderful, thanks for the quick reply!

**Comment by CamDavidsonPilon** [comment]
So I think the problem is in the dataset¹. You have rows that happen instaneously, example: 

<img width="357" alt="Screen Shot 2019-07-08 at 10 31 20 PM" src="https://user-images.githubusercontent.com/884032/60855003-21920d00-a1d0-11e9-90de-a3978db356b0.png">

This should be represented, equivalently, as 

<img width="376" alt="Screen Shot 2019-07-08 at 10 32 35 PM" src="https://user-images.githubusercontent.com/884032/60855053-4edebb00-a1d0-11e9-97e7-f677f81de204.png">

When I made these corrections to your dataset, the program ran fine. Ex:
[gh_issue copy 2.xlsx](https://github.com/CamDavidsonPilon/lifelines/files/3370933/gh_issue.copy.2.xlsx)

I'm curious, did you use lifelines to generate this dataset? How could this have been more obvious to fix? 

¹ However lifelines should have a check for this.

**Comment by indiana-nikel** [comment]
Hi Cam,

This dataset is a subset of what I'm working with and the format was generated using both `to_long_format` and `add_covariate_to_timeline`.  Looking at the fix you propose, I believe the problem lies within this step. A large issue I ran into is that there are no "death" events, only a tapering off of activity. I'd "solved" this by implementing a floor of time and at that point (which is their last activity time + the floor) they would have "died". This is coded as another row in the dataset where nothing happens in that time period except for the "death" event. 

I've taken the two datasets used in the formatting step of this and stripped them down similarly to the one I've shown above: 
[df_base.xlsx](https://github.com/CamDavidsonPilon/lifelines/files/3374152/df_base.xlsx)
[df_cv.xlsx](https://github.com/CamDavidsonPilon/lifelines/files/3374153/df_cv.xlsx)

Here is the code that I use to create the dataset fed into the CoxTimeVaryingFitter() above:
```
from lifelines.utils import to_long_format, add_covariate_to_timeline

df_base = pd.read_excel("df_base.xlsx")
df_cv = pd.read_excel("df_cv.xlsx")

df_ctv = df_base.pipe(to_long_format, duration_col="Duration")\
    .pipe(add_covariate_to_timeline, df_cv, duration_col="Time", id_col="ID", event_col="Event", cumulative_sum=True)
df_ctv.head()
```

Here we see those rows of instantaneous events pop up. I think this may also be in part to my misunderstanding of what sorts of timelines are expected to be fed into the `CoxTimeVaryingFitter()`.

Cheers,
Indiana

**Comment by CamDavidsonPilon** [comment]
Yes, I see the problem. I'll need to change some internal code to handle this. Thanks for the follow up!

**Comment by CamDavidsonPilon** [comment]
I need to think more about the following question:

> suppose we are recording a subject's measurements over time. One day, as we walk into the room, and the patient dies (say at time _t_). A moment later, at time _t+1_, we look at the subject's measurements, and record them. 

i) Are the data record at time _t+1_ "allowed" in the inference? 
ii) What if we had looked at measurements at time _t_ instead? 

It seems silly to discard this row, _but_ there are very valid cases where it makes sense to. Ex: if the measurement is heart rate, then at _t+1_ heart rate =0, and our inference will strongly suggest that "0 heart rate => death", which is reverse causality. 


What I'm trying to determine is if I should _keep_  observations that land on exactly the time of death, or should they be discarded (implicitly or explicitly by the user).

**Comment by indiana-nikel** [comment]
Looking at events that occur _exactly_ at the time of death think this would depend on what the event is. If it's a perfect correlation with "death" (heart rate = 0, clicks the 'Delete Account' button in a SaaS example), then we would have convergence problem (I ran into this in a similar but unrelated piece of work using the R package `survival`). Otherwise, that event might be a clear indicator of why patient/customer would die/leave. 

Does this question pertain to covariates changing over time or would it include categorical information to discard as well? 

This also raises the question of instantaneous events at the beginning of a timeline as well. In my example, I've added 1 so that there is no instantaneous events at t=0, all starting events should start at t=1. Would this suggest that it makes sense to do the same for a death event?
